cluster-autoscaler/cloudprovider/openstack/openstack_manager_heat.go

```diff
@@ -0,0 +1,299 @@
+	var cfg provider_os.Config
+	if configReader != nil {
+		if err := gcfg.ReadInto(&cfg, configReader); err != nil {
+			klog.Errorf("Couldn't read config: %v", err)
+			return nil, err
+		}
+	}
+
+	if opts.ClusterName == "" {
+		klog.Fatalf("The cluster-name parameter must be set")
+	}
+
+	authOpts := toAuthOptsExt(cfg)
+
+	provider, err := openstack.NewClient(cfg.Global.AuthURL)
+	if err != nil {
+		return nil, fmt.Errorf("could not authenticate client: %v", err)
+	}
+
+	if cfg.Global.CAFile != "" {
+		roots, err := certutil.NewPool(cfg.Global.CAFile)
+		if err != nil {
+			return nil, err
+		}
+		config := &tls.Config{}
+		config.RootCAs = roots
+		provider.HTTPClient.Transport = netutil.SetOldTransportDefaults(&http.Transport{TLSClientConfig: config})
+
+	}
+
+	err = openstack.AuthenticateV3(provider, authOpts, gophercloud.EndpointOpts{})
+	if err != nil {
+		return nil, fmt.Errorf("could not authenticate: %v", err)
+	}
+
+	clusterClient, err := openstack.NewContainerInfraV1(provider, gophercloud.EndpointOpts{Type: "container-infra", Name: "magnum", Region: cfg.Global.Region})
+	if err != nil {
+		return nil, fmt.Errorf("could not create container-infra client: %v", err)
+	}
+
+	heatClient, err := openstack.NewOrchestrationV1(provider, gophercloud.EndpointOpts{Type: "orchestration", Name: "heat", Region: cfg.Global.Region})
+	if err != nil {
+		return nil, fmt.Errorf("could not create orchestration client: %v", err)
+	}
+
+	manager := OpenstackManagerHeat{
+		clusterClient: clusterClient,
+		clusterName:   opts.ClusterName,
+		heatClient:    heatClient,
+		waitTimeStep:  waitForStatusTimeStep,
+	}
+
+	manager.stackID, err = manager.GetStackID()
+	if err != nil {
+		return nil, fmt.Errorf("could not store stack ID on manager: %v", err)
+	}
+	manager.stackName, err = manager.GetStackName()
+	if err != nil {
+		return nil, fmt.Errorf("could not store stack name on manager: %v", err)
+	}
+
+	return &manager, nil
+}
+
+// NodeGroupSize gets the current cluster size as reported by magnum.
+// The nodegroup argument is ignored as this implementation of OpenstackManager
+// assumes that only a single node group exists.
+func (osm *OpenstackManagerHeat) NodeGroupSize(nodegroup string) (int, error) {
+	cluster, err := clusters.Get(osm.clusterClient, osm.clusterName).Extract()
+	if err != nil {
+		return 0, fmt.Errorf("could not get cluster: %v", err)
+	}
+	return cluster.NodeCount, nil
+}
+
+// UpdateNodeCount replaces the cluster node_count in magnum.
+func (osm *OpenstackManagerHeat) UpdateNodeCount(nodegroup string, nodes int) error {
+	updateOpts := []clusters.UpdateOptsBuilder{
+		UpdateOptsInt{Op: clusters.ReplaceOp, Path: "/node_count", Value: nodes},
+	}
+	_, err := clusters.Update(osm.clusterClient, osm.clusterName, updateOpts).Extract()
+	if err != nil {
+		return fmt.Errorf("could not update cluster: %v", err)
+	}
+	return nil
+}
+
+// GetNodes currently returns nothing and this seems to work fine.
```

Comment:
From what I have seen, the interface method that calls this (nodeGroup.Nodes()) is only called when attempting to resolve unregistered nodes ([clusterstate.go](https://github.com/kubernetes/autoscaler/blob/cluster-autoscaler-release-1.13/cluster-autoscaler/clusterstate/clusterstate.go#L888)).  By not returning anything, the CA is never able to build the list of unregistered nodes.

Nodes are unregistered when they first come up, but also in certain error conditions (like a scheduling failure, or bootstrapping failure).  If the unregistered duration exceeds the "max-node-provision-time" flag, the CA attempts to scale out the node.

Our implementation involves hitting Nova to figure out which instances exist on the Openstack side, vs the nodes that exist on the kubernetes side.  If Openstack reports a node that k8s does not know about, then this presents the unregistered node condition.